securedrop/tests/test_encryption.py

```diff
@@ -0,0 +1,355 @@
+            gpg_key_dir=gpg_key_dir, journalist_key_fingerprint=journalist_key_fingerprint
+        )
+
+        # When trying to delete the journalist key via the encryption manager
+        # It fails
+        with pytest.raises(GpgKeyNotFoundError):
+            encryption_mgr.delete_source_key_pair(journalist_key_fingerprint)
+
+    def test_delete_source_key_pair_pinentry_status_is_handled(
+        self, source_app, test_source, mocker, capsys
+    ):
+        """
+        Regression test for https://github.com/freedomofpress/securedrop/issues/4294
+        """
+        # Given a source user with a key pair in the default encryption manager
+        source_user = test_source["source_user"]
+        encryption_mgr = EncryptionManager.get_default()
+
+        # And a gpg binary that will trigger the issue described in #4294
+        mocker.patch(
+            "pretty_bad_protocol._util._separate_keyword",
+            return_value=("PINENTRY_LAUNCHED", "does not matter"),
+        )
+
+        # When using the encryption manager to delete this source user's key pair
+        # It succeeds
+        encryption_mgr.delete_source_key_pair(source_user.filesystem_id)
+
+        # And the user's key information can no longer be retrieved
+        with pytest.raises(GpgKeyNotFoundError):
+            encryption_mgr.get_source_key_fingerprint(source_user.filesystem_id)
+
+        # And the bug fix was properly triggered
+        captured = capsys.readouterr()
+        assert "ValueError: Unknown status message: 'PINENTRY_LAUNCHED'" not in captured.err
+
+    def test_encrypt_source_message(self, setup_journalist_key_and_gpg_folder, tmp_path):
+        # Given an encryption manager
+        journalist_key_fingerprint, gpg_key_dir = setup_journalist_key_and_gpg_folder
+        encryption_mgr = EncryptionManager(
+            gpg_key_dir=gpg_key_dir, journalist_key_fingerprint=journalist_key_fingerprint
+        )
+
+        # And a message to be submitted by a source
+        message = "s3cr3t message"
+
+        # When the source tries to encrypt the message
+        # It succeeds
+        encrypted_message_path = tmp_path / "message.gpg"
+        encryption_mgr.encrypt_source_message(
+            message_in=message, encrypted_message_path_out=encrypted_message_path
+        )
+
+        # And the output file contains the encrypted data
+        encrypted_message = encrypted_message_path.read_bytes()
+        assert encrypted_message
+
+        # And the journalist is able to decrypt the message
+        with import_journalist_private_key(encryption_mgr):
+            decrypted_message = encryption_mgr._gpg.decrypt(encrypted_message).data
+        assert decrypted_message.decode() == message
+
+        # And the source or anyone else is NOT able to decrypt the message
+        # For GPG 2.1+, a non-null passphrase _must_ be passed to decrypt()
+        assert not encryption_mgr._gpg.decrypt(encrypted_message, passphrase="test 123").ok
```

Comment:
Example of a new test that ensures that only the right user/key can decrypt some data.